### PR TITLE
add docs for x-ms-header-collection-prefix

### PR DESCRIPTION
**Pull request by iscai-msft**

### CODE REVIEW

File: docs/extensions/readme.md [hunk 1]
```diff
@@ -20,15 +20,17 @@ The following documents describes AutoRest specific vendor extensions for [OpenA
 * [x-ms-error-response](#x-ms-error-response) - indicates whether the response status code should be treated as an error response
 * [x-ms-text](#x-ms-text) - XML helper for scenario with attributes and text on the same XML node
 * [x-ms-client-default](#x-ms-client-default) - Set default value for properties and parameters
+* [x-ms-pageable](#x-ms-pageable) - allows paging through lists of data.
```
Comment (iscai-msft) [inline docs/extensions/readme.md:23]: moved some of these extensions from the `Microsoft Azure Extensions (available in most generators only when using `--azure-arm`)` section. Def could've missed some extensions in that section though

File: docs/extensions/readme.md [hunk 2]
```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
...
+        "type": "string",
+        "x-ms-client-name": "Metadata",
+        "x-ms-header-collection-prefix": "x-ms-meta-"
+      }
+    }
+  }
+}
+```
+
+This response header parameter will be forced to be a dictionary schema. Only entries with prefix `x-ms-meta-` will be returned to users,
```
Comment (timotheeguerin) [inline docs/extensions/readme.md:1220]: Just so I completely understand. 
Let say I have a header `foo` defined as a string. Then in the generated it should provide a way to give a string value for that header as expected.
Now however if I add the `x-ms-header-collection-prefix` property it will automatically change the type of that header from string to `Dict<string, string>`?

File: docs/extensions/readme.md [hunk 3]
```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
...
+"parameters": [
+  "x-ms-meta": {
+    "type": "string",
+    "x-ms-client-name": "Metadata",
+    "x-ms-header-collection-prefix": "x-ms-meta-"
+  }
+]
+```
+This request parameter will be forced to be a dictionary schema, and all keys in this dictionary will be prefixed with `x-ms-meta-`.
+So, if you input `{"key": "value"}` through this parameter, `{"x-ms-meta-key": "value"}` is what reaches the service.
```
Comment (pakrym) [inline docs/extensions/readme.md:1201]: >`{"x-ms-meta-key": "value"}`


It would be a header, illustrating it with JSON might be a bit confusing.

File: docs/extensions/readme.md [hunk 4]
```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
...
+
+Additionally, applying this extension to a schema forces the schema to become a dictionary.
+
+**Schema**: `string`. Name of the prefix you want to append / filter by. A common value for storage libraries is `x-ms-meta-`.
+
+**Request Example**:
+
+```json
+"parameters": [
+  "x-ms-meta": {
```
Comment (pakrym) [inline docs/extensions/readme.md:1193]: nit: needs an object definition to be a valid JSON

**Review by pakrym — APPROVED** [review]

File: docs/extensions/readme.md [hunk 5]
```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
...
+    "type": "string",
+    "x-ms-client-name": "Metadata",
+    "x-ms-header-collection-prefix": "x-ms-meta-"
+  }
+]
+```
+This request parameter will be forced to be a dictionary schema, and all keys in this dictionary will be prefixed with `x-ms-meta-`.
```
Comment (lmazuel) [inline docs/extensions/readme.md:1200]: The `type` declared in the Swagger becomes the value type of the dict I guess?

File: docs/extensions/readme.md [hunk 6]
```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
...
+      }
+    }
+  }
+}
+```
+
+This response header parameter will be forced to be a dictionary schema. Only entries with prefix `x-ms-meta-` will be returned to users,
+and this prefix will be stripped before be returned to users. So if the response from the service is
+
+```json
```
Comment (lmazuel) [inline docs/extensions/readme.md:1226]: I don't think this example helps, since I believe this applies only to Headers, so having an example in JSON it confusing

File: docs/extensions/readme.md (hunk 4 above)
Comment (iscai-msft) [reply docs/extensions/readme.md]: totally defied this one wrong, fixed it. thanks!

File: docs/extensions/readme.md (hunk 5 above)
Comment (pakrym) [reply docs/extensions/readme.md:1203]: Can we define it for strings only? I think this is the only type currently in-use and would simplify the implementation.

File: docs/extensions/readme.md (hunk 2 above)
Comment (pakrym) [reply docs/extensions/readme.md:1223]: Correct

Comment (iscai-msft) [reply docs/extensions/readme.md:1223]: basically if you have a header parameter, even if it's schema is defined to be of type string, m4 should model it as a string dictioniary instead

this is based off the issue @pakrym opened for this https://github.com/Azure/autorest/issues/3761

File: docs/extensions/readme.md (hunk 5 above)
Comment (iscai-msft) [reply docs/extensions/readme.md:1203]: I'm following what @pakrym brought up in his [issue](https://github.com/Azure/autorest/issues/3761), where it should alwas be a string dictionary.

@pakrym should we model the value type to follow from the type listed instead of hardcoding it to string?

Comment (pakrym) [reply docs/extensions/readme.md:1203]: If it was me I would throw an error for any usage of `x-ms-header-collection-prefix` with a non-string type to cut down on the scope of this feature.

File: docs/extensions/readme.md (hunk 2 above)
Comment (timotheeguerin) [reply docs/extensions/readme.md:1223]: yeah so regardless of the documentation, if modeler4 decide to change that automatically is that going to create a breaking change for language already doing it?
would it convert `string` (m4)-> `Dict<string, string>` (lang)-> `Dict<string, Dict<string, string>>`

File: docs/extensions/readme.md (hunk 5 above)
Comment (iscai-msft) [reply docs/extensions/readme.md:1203]: rip sorry @pakrym, both times my github didn't update with your answer so i just ended up repeating you

File: docs/extensions/readme.md [hunk 7]
```diff
@@ -1173,3 +1175,59 @@ An object with an optional property.
   }
 }
 ```
+
+## x-ms-header-collection-prefix
+Use `x-ms-header-collection-prefix` to deal with prefixes for headers. Has different behavior if it refers to a request header or a response header:
```
Comment (annatisch) [inline docs/extensions/readme.md:1180]: I would rephrase this a little -  it's about handling collections of arbitrary headers by distinguishing them using a specified prefix.

File: docs/extensions/readme.md (hunk 2 above)
Comment (pakrym) [reply docs/extensions/readme.md:1223]: We would need to make sure everyone is aware of the change and ready to react. I suspect C# and GO are the only two languages that would need to.

Comment (iscai-msft) [reply docs/extensions/readme.md:1223]: also @timotheeguerin , pthon is the only language generator that has GA'd, and we don't have support foer it yet, so not an issue in terms for breaking GA'd libraries

File: docs/extensions/readme.md (hunk 5 above)
Comment (lmazuel) [reply docs/extensions/readme.md:1203]: I vote @pakrym , let's document that string is the only acceptable type, until someone prooves otherwise

File: docs/extensions/readme.md (hunk 2 above)
Comment (annatisch) [reply docs/extensions/readme.md:1223]: If I'm following your question correctly - the purpose of this extension is that it's for undefined headers. The SDK/swagger has no knowledge of what the header names might be.
So yes - it would be a collection/dictionary of arbitrary data, that was gathered from headers marked with the requisite prefix.

File: docs/extensions/readme.md (hunk 5 above)
Comment (annatisch) [reply docs/extensions/readme.md:1203]: Yes - I would say string is the only option - it will be up to the end user (or maybe the SDK if they have prior knowledge) to further deserialize this content.

File: docs/extensions/readme.md (hunk 3 above)
Comment (iscai-msft) [reply docs/extensions/readme.md:1204]: hey @pakrym does changing it to this clear it up?

`{"key": "value"}` through this parameter, the header `x-ms-meta: {"x-ms-meta-key": "value"}` ...
this is my understanding of what happens to the header param, lmk if this is wrong

Comment (pakrym) [reply docs/extensions/readme.md:1204]: You can use HTTP request syntax to illustrate
```
GET /path HTTP/1.1
x-ms-meta-key: value
```

Comment (iscai-msft) [reply docs/extensions/readme.md:1204]: sweet thanks so much!

**Review by jhendrixMSFT — APPROVED** [review]

File: docs/extensions/readme.md (hunk 5 above)
Comment (timotheeguerin) [reply docs/extensions/readme.md:1203]: Ok, will then look at implementing this change in m4 and ping go and csharp when the change is ready.

**Review by lmazuel — APPROVED** [review]

**Review by daviwil — APPROVED** [review]
